### PR TITLE
KIP 396 ListOffsets

### PR DESCRIPTION
**Pull request by mahajanadhitya**

### CODE REVIEW

File: kafka/integration_test.go [hunk 1]
```diff
@@ -2684,6 +2685,115 @@ func (its *IntegrationTestSuite) TestAdminClient_UserScramCredentials() {
...
+	defer cancel()
+
+	requests := make(map[TopicPartition]int64)
+	goTopic := "topicname"
+
+	var topics []TopicSpecification
+	topics = append(topics, TopicSpecification{Topic: goTopic, NumPartitions: 1, ReplicationFactor: 1})
+	create_topic_result, create_topic_err := a.CreateTopics(ctx, topics)
+	if create_topic_err != nil {
+		t.Fatalf("Unable to create topic, error : %s", err)
```
Comment (milindl) [inline kafka/integration_test.go:2715]: Now that assertions are available, use them.
```
assert := its.Assert()
....
....
assert.Nil(err, "CreateTopics should succeed")
```
and other methods on assert.
Everywhere, it's better readability and we will use this library going forward

**Review by milindl — CHANGES_REQUESTED** [review]
Some initial comments for Round 1. 
Mostly addresses the code, some comments on test, and example is not reviewed.
For the public API stuff and how OffsetSpec will be used, we can also discuss internally with everyone to arrive at a consensus on all the bindings at once.

File: kafka/adminapi.go [hunk 2]
```diff
@@ -2734,6 +2734,109 @@ func (a *AdminClient) DescribeUserScramCredentials(
...
+//   - `ctx` - context with the maximum amount of time to block, or nil for
+//     indefinite.
+//   - `requests` - a map from TopicPartition to int, int holds either the OffsetSpec enum value or timestamp.
+//   - `options` - ListOffsetsAdminOption options.
+//
+// Returns a map from TopicPartition to ListOffsetsResultInfo.
+// Each TopicPartition's ListOffset call can have an individual error.
+func (a *AdminClient) ListOffsets(
+	ctx context.Context, requests map[TopicPartition]int64,
+	options ...ListOffsetsAdminOption) (result map[TopicPartition]ListOffsetsResultInfo, err error) {
```
[the remark this answers is not in the document]
Comment (mahajanadhitya) [reply kafka/adminapi.go:2769]: I will further discuss it with emanuelle.

File: kafka/integration_test.go [hunk 3]
```diff
@@ -2684,6 +2684,97 @@ func (its *IntegrationTestSuite) TestAdminClient_UserScramCredentials() {
...
+		Timestamp:      t3,
+	}, nil)
+
+	p.Flush(5 * 1000)
+
+	tp1 := TopicPartition{Topic: &goTopic, Partition: 0}
+	topicPartitionOffsets[tp1] = EarliestOffsetSpec
+	var results ListOffsetsResult
+	results, err = a.ListOffsets(ctx, topicPartitionOffsets, SetAdminIsolationLevel(ReadCommitted))
+	assert.Nil(err, "ListOffsets request failed.")
```
[the remark this answers is not in the document]
Comment (milindl) [reply kafka/integration_test.go:2747]: Similar kind of language should be used everywhere, but this is a nit

File: kafka/adminapi.go [hunk 4]
```diff
@@ -2734,6 +2774,94 @@ func (a *AdminClient) DescribeUserScramCredentials(
 	return result, nil
 }
 
+// cToListOffsetsResult converts a C
+// rd_kafka_ListOffsets_result_t to a Go ListOffsetsResult
+func cToListOffsetsResult(cRes *C.rd_kafka_ListOffsets_result_t) (result ListOffsetsResult) {
```
Comment (milindl) [inline kafka/adminapi.go:2779]: Move this up with other cTo... functions

File: kafka/adminoptions.go [hunk 5]
```diff
@@ -171,6 +171,47 @@ func SetAdminRequestTimeout(t time.Duration) (ao AdminOptionRequestTimeout) {
...
+		C.rd_kafka_AdminOptions_destroy(cOptions)
+		return newErrorFromCErrorDestroy(cError)
+
+	}
+
+	return nil
+
+}
+
+func SetAdminIsolationLevel(isolationLevel IsolationLevel) (ao AdminOptionIsolationLevel) {
```
Comment (milindl) [inline kafka/adminoptions.go:209]: Documentation comment above this missing

File: kafka/integration_test.go [hunk 6]
```diff
@@ -2684,6 +2684,97 @@ func (its *IntegrationTestSuite) TestAdminClient_UserScramCredentials() {
 	}
 }
 
+// Tests ListOffsets API which describes
+// the offset of a TopicPartiton corresponding to the OffsetSpec provided.
```
Comment (milindl) [inline kafka/integration_test.go:2688]: Partition*

^ spelling

File: kafka/adminoptions.go [hunk 7]
```diff
@@ -171,6 +171,47 @@ func SetAdminRequestTimeout(t time.Duration) (ao AdminOptionRequestTimeout) {
...
 }
 
+type IsolationLevel int
+
+const (
+	ReadUncommitted = IsolationLevel(C.RD_KAFKA_ISOLATION_LEVEL_READ_UNCOMMITTED)
+	ReadCommitted   = IsolationLevel(C.RD_KAFKA_ISOLATION_LEVEL_READ_COMMITTED)
+)
+
+type AdminOptionIsolationLevel struct {
```
Comment (milindl) [inline kafka/adminoptions.go:181]: Missing comment above this

Comment (milindl) [reply kafka/adminoptions.go:181]: See other stuff around this for how to write the comment

**Review by milindl — CHANGES_REQUESTED** [review]
Example is missing from README.md of the examples/ directory, and also the .gitgnore (see how the other examples are added)

I also checked valgrind for memory leaks in the example. No issues

Mostly, great work, thanks @mahajanadhitya !

File: kafka/adminoptions.go [hunk 8]
```diff
@@ -171,6 +171,57 @@ func SetAdminRequestTimeout(t time.Duration) (ao AdminOptionRequestTimeout) {
...
+
+	return nil
+
+}
+
+// SetAdminIsolationLevel sets the overall IsolationLevel for a request.
+//
+// Default: `ReadUncommitted`.
+//
+// Valid for all Admin API methods.
```
Comment (milindl) [inline kafka/adminoptions.go:218]:
```suggestion
// Valid for ListOffsets.
```

Comment (milindl) [reply kafka/adminoptions.go:218]: . It's valid only for ListOffsets

File: kafka/integration_test.go [hunk 9]
```diff
@@ -2684,6 +2684,93 @@ func (its *IntegrationTestSuite) TestAdminClient_UserScramCredentials() {
...
+		assert.Equal(info.Error.Code(), ErrNoError, "Error code should be 0.")
+		assert.Equal(info.Offset, int64(3), "Offset should be 3.")
+	}
+
+	topicPartitionOffsets[tp1] = OffsetSpec(MaxTimestampOffsetSpec)
+	results, err = a.ListOffsets(ctx, topicPartitionOffsets, SetAdminIsolationLevel(ReadCommitted))
+	assert.Nil(err, "ListOffsets should not fail.")
+
+	for _, info := range results.Results {
+		assert.Equal(info.Error.Code(), ErrNoError, "Error code should be 0.")
```
Comment (milindl) [inline kafka/integration_test.go:2764]: The message should be changed too, not just the assert. 
Like
```suggestion
		assert.Equal(info.Error.Code(), ErrNoError, "Error code should be ErrNoError.")
```

Comment (milindl) [reply kafka/integration_test.go:2764]: At all three places.

File: kafka/adminoptions.go [hunk 10]
```diff
@@ -171,6 +171,57 @@ func SetAdminRequestTimeout(t time.Duration) (ao AdminOptionRequestTimeout) {
...
+const (
+	ReadUncommitted = IsolationLevel(C.RD_KAFKA_ISOLATION_LEVEL_READ_UNCOMMITTED)
+	ReadCommitted   = IsolationLevel(C.RD_KAFKA_ISOLATION_LEVEL_READ_COMMITTED)
+)
+
+// AdminOptionIsolationLevel sets the overall request IsolationLevel.
+//
+// Default: `ReadUncommitted`.
+//
+// Valid for all Admin API methods.
```
Comment (milindl) [inline kafka/adminoptions.go:185]:
```suggestion
// Valid for ListOffsets.
```

**Review by milindl — COMMENTED** [review]
Minor comments on the addressal itself.
Almost ready to merge.

**Review by milindl — APPROVED** [review]
Approved, don't merge until librdkafka bundle is merged.
Good work @mahajanadhitya !

File: kafka/adminapi.go [hunk 11]
```diff
@@ -986,6 +993,36 @@ type AlterUserScramCredentialsResult struct {
...
+type ListOffsetsResultInfo struct {
+	Offset      int64
+	Timestamp   int64
+	LeaderEpoch int
+	Error       Error
+}
+
+// ListOffsetsResult holds the map of TopicPartition to ListOffsetsResultInfo for a request.
+type ListOffsetsResult struct {
+	Results map[TopicPartition]ListOffsetsResultInfo
```
Comment (emasab) [inline kafka/adminapi.go:1023]:
```suggestion
	ResultsInfos map[TopicPartition]ListOffsetsResultInfo
```

File: kafka/adminapi.go [hunk 12]
```diff
@@ -986,6 +993,36 @@ type AlterUserScramCredentialsResult struct {
...
+// NewOffsetSpecForTimestamp creates an OffsetSpec corresponding to the timestamp.
+func NewOffsetSpecForTimestamp(timestamp int64) OffsetSpec {
+	return OffsetSpec(timestamp)
+}
+
+// ListOffsetsResultInfo describes the result of ListOffsets request for a Topic Partition.
+type ListOffsetsResultInfo struct {
+	Offset      int64
+	Timestamp   int64
+	LeaderEpoch int
```
Comment (emasab) [inline kafka/adminapi.go:1017]: Should be nil if -1, as in TopicPartition
```suggestion
	LeaderEpoch *int32
```

File: kafka/adminapi.go [hunk 13]
```diff
@@ -1366,6 +1403,27 @@ func cToDescribeUserScramCredentialsResult(
...
 
+// cToListOffsetsResult converts a C
+// rd_kafka_ListOffsets_result_t to a Go ListOffsetsResult
+func cToListOffsetsResult(cRes *C.rd_kafka_ListOffsets_result_t) (result ListOffsetsResult) {
+	result = ListOffsetsResult{Results: make(map[TopicPartition]ListOffsetsResultInfo)}
+	var cPartitionCount C.size_t
+	cResultInfos := C.rd_kafka_ListOffsets_result_infos(cRes, &cPartitionCount)
+	for itr := 0; itr < int(cPartitionCount); itr++ {
+		cResultInfo := C.ListOffsetsResultInfo_by_idx(cResultInfos, cPartitionCount, C.size_t(itr))
+		Value := ListOffsetsResultInfo{}
```
Comment (emasab) [inline kafka/adminapi.go:1414]:
```suggestion
		resultInfo := ListOffsetsResultInfo{}
```

File: kafka/adminapi.go [hunk 14]
```diff
@@ -986,6 +993,36 @@ type AlterUserScramCredentialsResult struct {
...
+)
+
+// NewOffsetSpecForTimestamp creates an OffsetSpec corresponding to the timestamp.
+func NewOffsetSpecForTimestamp(timestamp int64) OffsetSpec {
+	return OffsetSpec(timestamp)
+}
+
+// ListOffsetsResultInfo describes the result of ListOffsets request for a Topic Partition.
+type ListOffsetsResultInfo struct {
+	Offset      int64
```
Comment (emasab) [inline kafka/adminapi.go:1015]:
```suggestion
	Offset      kafka.Offset
```

File: kafka/adminapi.go [hunk 15]
```diff
@@ -1366,6 +1403,27 @@ func cToDescribeUserScramCredentialsResult(
...
+	cResultInfos := C.rd_kafka_ListOffsets_result_infos(cRes, &cPartitionCount)
+	for itr := 0; itr < int(cPartitionCount); itr++ {
+		cResultInfo := C.ListOffsetsResultInfo_by_idx(cResultInfos, cPartitionCount, C.size_t(itr))
+		Value := ListOffsetsResultInfo{}
+		cPartition := C.rd_kafka_ListOffsetsResultInfo_topic_partition(cResultInfo)
+		Topic := C.GoString(cPartition.topic)
+		Partition := TopicPartition{Topic: &Topic, Partition: int32(cPartition.partition)}
+		Value.Offset = int64(cPartition.offset)
+		Value.Timestamp = int64(C.rd_kafka_ListOffsetsResultInfo_timestamp(cResultInfo))
+		Value.LeaderEpoch = -1
```
Comment (emasab) [inline kafka/adminapi.go:1420]:
```suggestion
		cLeaderEpoch := int32(C.rd_kafka_topic_partition_get_leader_epoch(cPartition))
		if cLeaderEpoch >= 0 {
			resultInfo.LeaderEpoch = &cLeaderEpoch
		}
```

File: kafka/adminoptions.go [hunk 16]
```diff
@@ -175,6 +175,60 @@ func SetAdminRequestTimeout(t time.Duration) (ao AdminOptionRequestTimeout) {
 	return ao
 }
 
+// IsolationLevel is a type which is used for AdminOptions to set the IsolationLevel.
+type IsolationLevel int
+
+const (
+	// ReadUncommitted IsolationLevel
+	ReadUncommitted = IsolationLevel(C.RD_KAFKA_ISOLATION_LEVEL_READ_UNCOMMITTED)
```
Comment (emasab) [inline kafka/adminoptions.go:183]: As other enums:
```suggestion
	IsolationLevelReadUncommitted = IsolationLevel(C.RD_KAFKA_ISOLATION_LEVEL_READ_UNCOMMITTED)
```

File: kafka/adminoptions.go [hunk 17]
```diff
@@ -175,6 +175,60 @@ func SetAdminRequestTimeout(t time.Duration) (ao AdminOptionRequestTimeout) {
...
+	// ReadCommitted IsolationLevel
+	ReadCommitted = IsolationLevel(C.RD_KAFKA_ISOLATION_LEVEL_READ_COMMITTED)
```
Comment (emasab) [inline kafka/adminoptions.go:185]:
```suggestion
	IsolationLevelReadCommitted = IsolationLevel(C.RD_KAFKA_ISOLATION_LEVEL_READ_COMMITTED)
```

File: kafka/adminapi.go [hunk 18]
```diff
@@ -1366,6 +1403,27 @@ func cToDescribeUserScramCredentialsResult(
...
+	result = ListOffsetsResult{Results: make(map[TopicPartition]ListOffsetsResultInfo)}
+	var cPartitionCount C.size_t
+	cResultInfos := C.rd_kafka_ListOffsets_result_infos(cRes, &cPartitionCount)
+	for itr := 0; itr < int(cPartitionCount); itr++ {
+		cResultInfo := C.ListOffsetsResultInfo_by_idx(cResultInfos, cPartitionCount, C.size_t(itr))
+		Value := ListOffsetsResultInfo{}
+		cPartition := C.rd_kafka_ListOffsetsResultInfo_topic_partition(cResultInfo)
+		Topic := C.GoString(cPartition.topic)
+		Partition := TopicPartition{Topic: &Topic, Partition: int32(cPartition.partition)}
+		Value.Offset = int64(cPartition.offset)
```
Comment (emasab) [inline kafka/adminapi.go:1418]:
```suggestion
		Value.Offset = kafka.Offset(cPartition.offset)
```

**Review by emasab — CHANGES_REQUESTED** [review]
Great job! Some small changes

**Comment by milindl** [comment]
Addressed comments and merged master.

File: kafka/adminapi.go (hunk 18 above)
Comment (milindl) [reply kafka/adminapi.go]: Just using Offset() here since the package is already kafka

File: kafka/adminapi.go (hunk 14 above)
Comment (milindl) [reply kafka/adminapi.go]: Just 'Offset'

**Review by emasab — APPROVED** [review]
LGTM!

**Pull request merged**